### ISSUE TITLE
Parser gap: anonymous invocation nested 34+ non-brace constructs deep is rejected when a later syntax error exists

### ISSUE THREAD

**Issue by ss-o** [z-shell/zsh-lint#267]
## Summary

`prefixEndsWithAnonymousFunction` (added by PR #264's review remediation) validates an anonymous-function invocation candidate by reparsing the source prefix ending at the candidate's `}` and, on each "incomplete construct" error, appending the closer the error names and retrying. The retry is bounded by `maxClosers = 32 + braceOpenerCount(prefix)`: brace nesting is derived from the prefix, but nesting inside non-brace constructs (`if`/`fi`, `for`/`done`, `while`, `until`, `case`/`esac`, `select`, `repeat`, subshells, open quotes) is covered only by the fixed 32.

A genuine invocation inside 34 nested `if` blocks has only its own `{` in the prefix, so `maxClosers` is 33 while validation needs 34 `fi` tokens. When a later, separate blocker in the same file sends `parseAnonymousFunctionArgs` down its error-return path, the genuine candidate is rejected and its stale `seedErr` is reported instead of the real later error. A file with no later blocker is unaffected: the success path binds candidates from the whole-file parse and never enters this validation.

For this shape, depth 33 is the last depth that works and 34 is the first false rejection (the invocation's own `{` contributes 1 to the bound, so a body with more braces shifts the threshold). Brace nesting has no such cap since `braceOpenerCount` was added (PR #264, round 4 and round 5 review remediation).

## Minimized script

```zsh
if true; then
if true; then
# ... 34 `if true; then` lines in total ...
() { x; } y
fi
fi
# ... 34 `fi` lines in total ...
)
```

Generated with `zsh -fc 'repeat 34 print "if true; then"; print "() { x; } y"; repeat 34 print fi; print ")"'`. The trailing `)` is the later blocker on line 70; without it the file is valid and both native Zsh and zsh-lint accept it.

## `zsh -f -n` on that script

fails, as it should, at the trailing `)`: `` probe-nested-if-34-blocker.zsh:70: parse error near `)' ``. The same file without the trailing `)` passes (no output, exit 0).

## zsh-lint output

```
probe-nested-if-34-blocker.zsh:35:11: statements must be separated by &, ; or a newline
```

Line 35 column 11 is the genuine invocation's `y`, the candidate's stale `seedErr`. Expected: `70:1`, the trailing `)`, which is what the same binary reports for the identical file at depth 33 (`68:1`).

## Mechanism

`prefixEndsWithAnonymousFunction` in `internal/parse/anonymous_function_args.go` (PR #264 branch `bug-255` at `b205ad5`, line 166): `maxClosers := 32 + braceOpenerCount(prefix)`. The loop returns `false` once `appended >= maxClosers` even when every retry so far produced a genuine incompleteness error naming the next `fi`. The caller in `parseAnonymousFunctionArgs` then returns the rejected candidate's `seedErr`.

The fixed 32 exists for two reasons, both stated in the function's doc comment: `closingToken` matches any error ending in a backtick-quoted token, not only genuine incompleteness, and adapter rewrites on a truncated prefix can in principle ping-pong between requested closers instead of converging, so a hard cap is needed regardless of nesting. Deriving non-brace depth was an explicit scope decision in PR #264, not a claim that it cannot be done.

Copilot's automated review of `b205ad5` suggested two designs. Either one needs a first step that has not been done yet: establish what `Pos` means for every incompleteness shape `closingToken` can match. In the pinned `mvdan.cc/sh/v3@v3.14.1` (`syntax/parser.go`), `matchingErr` and `quoteErr` report the opener's position (`lpos`) and `stmtEnd` reports the enclosing statement's start (`n.Pos()`), so for those three a genuine nesting level yields a new position on each retry while the text stays the same (the text embeds the current token, `EOF` on every retry), and an exact `(Text, Pos)` repeat would mean no progress; `followErr` takes a caller-supplied position and needs the same check per call site. This repository's own adapters can also rewrite or rebase errors on the truncated prefix before `closingToken` sees them, so the probe has to cover the adapter chain, not only mvdan/sh.

1. Make termination depend on parser progress or cycle detection: stop when a retry reproduces an error already seen, with a large backstop cap that only pathological input reaches.
2. Derive a safe bound for non-brace constructs the same way `braceOpenerCount` does for braces: count opener bytes and keywords (`(`, `if`, `for`, `while`, `until`, `case`, `select`, `repeat`, `function`, quote characters) as a one-directional over-count. Missing any opener kind is a silent under-count, so the list must be checked against every construct the parser (and every compat adapter) can leave open.

On `main` at `6f06274`, before PR #264, the depth-34 file reports the same stale `35:11`, and the depth-2 and depth-33 variants report `3:11` and `34:11`: the stale error appears at every depth on `main`, which is the original #255 defect. PR #264 fixes it for brace nesting at any depth and for non-brace nesting up to 33. This issue is the residual band at depth 34 and beyond.

## Classification

zshmisc, Complex Commands (`if list then list ... fi`, nesting) and Functions (anonymous functions, "the function is executed immediately" with arguments). No manual limit on nesting depth applies; this is a validator bound, not a language rule.

## Where it occurs

Found by Copilot's automated review of PR #264 (suppressed comment on commit `b205ad5`, `internal/parse/anonymous_function_args.go:166`). Not observed in the reference corpus (`docs/project/corpus.md`, 18 files, all ok on `b205ad5`); 34 nested non-brace constructs around an anonymous invocation, in a file that also has a later syntax error, is a pathological shape.

## zsh-lint version

`main` @ `5d48f4a` (the squash merge of PR #264; reproduced on its head commit `b205ad5`, reachable through the PR). `main` @ `6f06274`, before that merge, shows the same symptom at every depth via the original #255 mechanism.

## Acceptance criteria

- First step: a written summary (in the fixing PR) of the `Pos` semantics of each mvdan/sh error constructor `closingToken` can match, so the chosen termination criterion is grounded in the parser's behavior rather than assumed.
- A `gap-<id>-deep-nested-if-anonymous-args.zsh` fixture (34 nested `if`, the invocation, 34 `fi`) passes `zsh -f -n` and parses cleanly; the same fixture with a trailing `)` reports the `)` line, not the invocation's line, on the fixed parser.
- `prefixEndsWithAnonymousFunction`'s termination no longer depends on a fixed allowance for non-brace nesting: either it detects non-progress across retries or its bound is derived from every opener kind the parser can leave open.
- `TestAnonymousFunctionArgsRetryHandlesDeepNesting`, `TestAnonymousFunctionArgsRetryHandlesNestingPastFixedBound`, `TestAnonymousFunctionArgsRetryHandlesNonBraceEnclosingConstructs`, and `TestAnonymousFunctionArgsRetryHandlesCommentBracesBeforeCandidate` pass unchanged.
- `zi.zsh` (`z-shell/zi`, `origin/next`, `bcb20c1`) still reports `1457:3` at ~0.26s, no regression: that file exercises the validation path, so any change to the retry's termination shows up there first.

## References

- #255 and PR #264: the fix that introduced the validation retry and its bound; the PR body's Round 2, Round 4, and Round 5 sections record why the fixed component exists and why brace depth is derived but non-brace depth is not.
- #263, #265, #266: other parser gaps filed from the same investigation.
